File: quicksort.py
import time
import statistics
#Loon suvalise listi, n on listi pikkus, k on suvaliste arvude vahemik
import random
randomList = []
n = 1000000
k = 100000
for i in range(n):
    randomList.append(random.randint(0,k))

def kiirsortimine(arr):
    #valin median of 3 strateegiaga tugielemendi
    if len(arr) >= 3:
        pivot = statistics.median([arr[0], arr[(len(arr)-1)//2], arr[len(arr)-1]]) 
        arr.remove(pivot)
        arr1 = []
        arr2 = []
        for i in arr:
            
            if i < pivot:
                arr1.append(i)
                 
            else:
                arr2.append(i)
        

        arr = kiirsortimine(arr1) + [pivot] + kiirsortimine(arr2) 
    elif len(arr) == 2:
        if arr[0] > arr[1]:
            arr[0], arr[1] = arr[1], arr[0]
    return(arr)
    
#print(randomList)
start = time.time()
kiirsortimine(randomList)
print(time.time()-start)

File: quicksort.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <array>
#include <chrono>
#include <vector>
#include <thread>

using namespace std;
using namespace std::chrono;

void printArray(vector<int> arr, int n)
{
    int i;
    for (i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
}

vector<int> joinArrays(vector<int> arr1, vector<int> arr2, int pivot) {
    vector<int> arr;
    arr.insert(arr.end(), arr1.begin(), arr1.end());
    arr.push_back(pivot);
    arr.insert(arr.end(), arr2.begin(), arr2.end());
    return arr;
}

vector<int> kiirSortimine(vector<int> &arr) {
int len = arr.size();
if (len >= 3) {
    int median = max(min(arr[0], arr[floor((len-1) /2)]), min(max(arr[0],arr[floor((len-1) /2)]), arr[len-1]));
    //cout<<"median: " << median << endl;
    auto it = find(arr.begin(), arr.end(), median);
    arr.erase(it);
    vector<int> arr1;
    vector<int> arr2;
    //cout << "arr: ";
    //printArray(arr, arr.size());
    for (auto &i : arr) {
        if (i<median) {
            arr1.push_back(i);
        }
        else {
            arr2.push_back(i);
        }
    }
        
        arr.clear();
        //cout<< "arr1: ";
        //printArray(arr1, arr1.size());
        //cout<< "arr2: ";
        //printArray(arr2, arr2.size());
        arr1 = kiirSortimine(arr1);
        arr2 = kiirSortimine(arr2);
        

        arr = joinArrays(arr1, arr2, median);

        //printArray(arr, arr.size());
       // this_thread::sleep_for(chrono::milliseconds(500));
        }
else if (len == 2) {
    if (arr[0] > arr[1]) {
        swap(arr[0], arr[1]);
    }
    }  
    return arr;
}

int main() {
    
    const int n = 1000000;
    const int k = 1000000;
    vector<int> vect;

    srand(time(NULL));
    for (int i = 0; i < n; ++i){
        
        vect.push_back(rand() %k) ;
    }
    //printArray(vect, n);

    auto start = high_resolution_clock::now();
    kiirSortimine(vect);
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    cout << endl;
    
    cout << duration.count() << " microseconds";
    //printArray(vect, n);
    return 0;
}